File: spvmoduledash/src/main/java/com/mycelium/spvmodule/dash/providers/BlockchainContentProvider.java
package com.mycelium.spvmodule.dash.providers;

import android.content.ContentProvider;
import android.content.ContentUris;
import android.content.ContentValues;
import android.content.Context;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.database.sqlite.SQLiteQueryBuilder;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.mycelium.spvmodule.dash.providers.DB.Addresses;
import com.mycelium.spvmodule.dash.providers.DB.Transactions;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class BlockchainContentProvider extends ContentProvider {
    public static final String AUTHORITY = "com.mycelium.spvmodule.dash.providers.BlockchainContentProvider";

    private static final UriMatcher URI_MATCHER;

    private static final String ADDRESSES_TABLE_NAME = "addresses";
    private static final String TRANSACTIONS_TABLE_NAME = "transactions";

    private static final int ADDRESS_LIST = 1;
    private static final int ADDRESS_ID = 2;
    private static final int TRANSACTION_LIST = 3;
    private static final int TRANSACTION_ID = 4;

    private static final Logger log = LoggerFactory.getLogger(BlockchainContentProvider.class);

    private static class DatabaseHelper extends SQLiteOpenHelper {
        private static final String DATABASE_NAME = "spvModuleDash.db";
        private static final int DATABASE_VERSION = 2;

        DatabaseHelper(Context context) {
            super(context, DATABASE_NAME, null, DATABASE_VERSION);
        }

        @Override
        public void onCreate(SQLiteDatabase db) {
            db.execSQL("DROP TABLE IF EXISTS " + ADDRESSES_TABLE_NAME + ";");
            db.execSQL("CREATE TABLE " + ADDRESSES_TABLE_NAME +
                    " (" + Addresses.ADDRESS + " TEXT PRIMARY KEY," +
                    Addresses.CREATION_DATE + " INTEGER," +
                    Addresses.SYNCED_TO_BLOCK + " INTEGER," +
                    Addresses.ACCOUNT + " TEXT);");
            db.execSQL("DROP TABLE IF EXISTS " + TRANSACTIONS_TABLE_NAME + ";");
            db.execSQL("CREATE TABLE " + TRANSACTIONS_TABLE_NAME +
                    " (" + Transactions.TRANSACTION_ID + " TEXT PRIMARY KEY," +
                    Transactions.TRANSACTION + " BLOB," +
                    Transactions.INCLUDED_IN_BLOCK + " INTEGER," +
                    Transactions.ACCOUNT + " TEXT);");
        }

        @Override
        public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
            log.warn("Upgrading database from version " + oldVersion + " to " + newVersion);
            if (oldVersion < 2) {
                onCreate(db);
            }
        }
    }

    private DatabaseHelper dbHelper;

    static {
        URI_MATCHER = new UriMatcher(UriMatcher.NO_MATCH);
        URI_MATCHER.addURI(AUTHORITY, ADDRESSES_TABLE_NAME, ADDRESS_LIST);
        URI_MATCHER.addURI(AUTHORITY, ADDRESSES_TABLE_NAME + "/#", ADDRESS_ID);
        URI_MATCHER.addURI(AUTHORITY, TRANSACTIONS_TABLE_NAME, TRANSACTION_LIST);
        URI_MATCHER.addURI(AUTHORITY, TRANSACTIONS_TABLE_NAME + "/#", TRANSACTION_ID);
    }

    @Override
    public boolean onCreate() {
        dbHelper = new DatabaseHelper(getContext());
        return true;
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, String[] projection, String selection, String[] selectionArgs, String sortOrder) {
        SQLiteQueryBuilder qb = new SQLiteQueryBuilder();
        int match = URI_MATCHER.match(uri);
        switch (match) {
            case ADDRESS_LIST:
            case TRANSACTION_LIST:
                break;
            case ADDRESS_ID:
            case TRANSACTION_ID:
                selection = selection + " _id=" + uri.getLastPathSegment();
                break;
            default:
                throw new IllegalArgumentException("Unknown URI " + uri);
        }
        qb.setTables(getTableFromMatch(match));
        SQLiteDatabase db = dbHelper.getReadableDatabase();
        Cursor cursor = qb.query(db, projection, selection, selectionArgs, null, null, sortOrder);
        //noinspection ConstantConditions
        cursor.setNotificationUri(getContext().getContentResolver(), uri);
        return cursor;
    }

    private static String getTableFromMatch(int match) {
        switch (match) {
            case ADDRESS_LIST:
            case ADDRESS_ID:
                return ADDRESSES_TABLE_NAME;
            case TRANSACTION_LIST:
            case TRANSACTION_ID:
                return TRANSACTIONS_TABLE_NAME;
            default:
                throw new IllegalArgumentException("Unknown match " + match);
        }
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        switch (URI_MATCHER.match(uri)) {
            case ADDRESS_LIST:
                return Addresses.CONTENT_TYPE;
            case ADDRESS_ID:
                return Addresses.CONTENT_ITEM_TYPE;
            case TRANSACTION_LIST:
                return Transactions.CONTENT_TYPE;
            case TRANSACTION_ID:
                return Transactions.CONTENT_ITEM_TYPE;
            default:
                throw new IllegalArgumentException("Unknown URI " + uri);
        }
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, ContentValues values) {
        // TODO: 9/14/16 validate data here
        if (values == null || values.size() < 1) {
            throw new IllegalArgumentException("Can't insert empty row!");
        }
        Uri contentUri;
        int match = URI_MATCHER.match(uri);
        switch (match) {
            case ADDRESS_LIST:
                contentUri = Addresses.CONTENT_URI;
                break;
            case TRANSACTION_LIST:
                contentUri = Transactions.CONTENT_URI;
                break;
            default:
                throw new IllegalArgumentException("Unknown URI " + uri);
        }
        SQLiteDatabase db = dbHelper.getWritableDatabase();
        long rowId = db.insert(getTableFromMatch(match), null, values);
        if (rowId > 0) {
            Uri noteUri = ContentUris.withAppendedId(contentUri, rowId);
            notifyChange(noteUri);
            return noteUri;
        }
        throw new SQLException("Failed to insert row into " + uri);
    }

    @Override
    public int delete(@NonNull Uri uri, String where, String[] whereArgs) {
        // TODO: 9/14/16 delete should not allow to delete transactions of watched addresses
        // TODO: 9/14/16 deleting addresses should delete their transactions
        SQLiteDatabase db = dbHelper.getWritableDatabase();
        int match = URI_MATCHER.match(uri);
        switch (match) {
            case ADDRESS_LIST:
            case TRANSACTION_LIST:
                break;
            case ADDRESS_ID:
            case TRANSACTION_ID:
                where = where + " _id=" + uri.getLastPathSegment();
                break;
            default:
                throw new IllegalArgumentException("Unknown URI " + uri);
        }
        int count = db.delete(getTableFromMatch(match), where, whereArgs);
        if (count > 0) {
            notifyChange(uri);
        }
        return count;
    }

    @Override
    public int update(@NonNull Uri uri, ContentValues values, String where, String[] whereArgs) {
        // TODO: 9/14/16 updates should be limited to very few things
        SQLiteDatabase db = dbHelper.getWritableDatabase();
        int match = URI_MATCHER.match(uri);
        int count = db.update(getTableFromMatch(match), values, where, whereArgs);
        if (count > 0) {
            notifyChange(uri);
        }
        return count;
    }

    private void notifyChange(Uri uri) {
        //crashing if context or contentResolver are null is appropriate here
        //noinspection ConstantConditions
        getContext().getContentResolver().notifyChange(uri, null);
    }
}

File: spvmodule/src/main/java/com/mycelium/spvmodule/providers/BlockchainContentProvider.kt
package com.mycelium.spvmodule.providers

import android.content.ContentProvider
import android.content.ContentUris
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.database.SQLException
import android.database.sqlite.SQLiteQueryBuilder
import android.net.Uri

import com.mycelium.spvmodule.providers.data.DatabaseHelper
import com.mycelium.modularizationtools.CommunicationManager
import com.mycelium.spvmodule.providers.BlockchainContract.Address
import com.mycelium.spvmodule.providers.BlockchainContract.Transaction
import com.mycelium.spvmodule.providers.BlockchainContract.TransactionOutput

class BlockchainContentProvider : ContentProvider() {
    private var dbHelper: DatabaseHelper? = null
    var communicationManager: CommunicationManager? = null

    override fun onCreate(): Boolean {
        dbHelper = DatabaseHelper(context)
        communicationManager = CommunicationManager.getInstance(context)
        return true
    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?, selectionArgs: Array<String>?, sortOrder: String?): Cursor? {
        var sel = selection
        checkSignature(callingPackage)
        val qb = SQLiteQueryBuilder()
        val match = URI_MATCHER.match(uri)
        when (match) {
            ADDRESS_LIST, TRANSACTION_LIST -> {
            }
            ADDRESS_ID, TRANSACTION_ID -> sel = sel + " _id=" + uri.lastPathSegment
            else -> throw IllegalArgumentException("Unknown URI " + uri) as Throwable
        }
        qb.tables = getTableFromMatch(match)
        val db = dbHelper!!.readableDatabase
        val cursor = qb.query(db, projection, sel, selectionArgs, null, null, sortOrder)
        //noinspection ConstantConditions
        cursor.setNotificationUri(context!!.contentResolver, uri)
        return cursor
    }

    override fun getType(uri: Uri): String? {
        checkSignature(callingPackage)
        return when (URI_MATCHER.match(uri)) {
            ADDRESS_LIST -> Address.CONTENT_TYPE
            ADDRESS_ID -> Address.CONTENT_ITEM_TYPE
            TRANSACTION_LIST -> Transaction.CONTENT_TYPE
            TRANSACTION_ID -> Transaction.CONTENT_ITEM_TYPE
            TRANSACTION_OUTPUT_LIST -> TransactionOutput.CONTENT_TYPE
            TRANSACTION_OUTPUT_ID -> TransactionOutput.CONTENT_ITEM_TYPE
            else -> throw IllegalArgumentException("Unknown URI " + uri)
        }
    }

    /**
     * Unconfirmed transactions can be inserted for broadcast using the TRANSACTION_LIST uri.
     *
     * Unmonitored addresses can be inserted for monitoring using the ADDRESS_LIST uri.
     */
    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        checkSignature(callingPackage)
        // TODO: 9/14/16 validate data here
        if (values == null || values.size() < 1) {
            throw IllegalArgumentException("Can't insert empty row!")
        }
        val match = URI_MATCHER.match(uri)
        val contentUri = when (match) {
            ADDRESS_LIST -> Address.CONTENT_URI(context.packageName)
            TRANSACTION_LIST -> Transaction.CONTENT_URI(context.packageName)
            TRANSACTION_OUTPUT_LIST -> TransactionOutput.CONTENT_URI(context.packageName)
            else -> throw IllegalArgumentException("Unknown URI " + uri)
        }
        val db = dbHelper!!.writableDatabase
        val rowId = db.insert(getTableFromMatch(match), null, values)
        if (rowId > 0) {
            val noteUri = ContentUris.withAppendedId(contentUri, rowId)
            notifyChange(noteUri)
            return noteUri
        }
        throw SQLException("Failed to insert row into " + uri)
    }

    override fun delete(uri: Uri, where: String?, whereArgs: Array<String>?): Int {
        var sel = where
        checkSignature(callingPackage)
        // TODO: 9/14/16 delete should not allow to delete transactions of watched addresses
        // TODO: 9/14/16 deleting addresses should delete their transactions
        val db = dbHelper!!.writableDatabase
        val match = URI_MATCHER.match(uri)
        when (match) {
            ADDRESS_LIST, TRANSACTION_LIST -> {
            }
            ADDRESS_ID, TRANSACTION_ID -> sel = sel + " _id=" + uri.lastPathSegment
            else -> throw IllegalArgumentException("Unknown URI " + uri)
        }
        val count = db.delete(getTableFromMatch(match), sel, whereArgs)
        if (count > 0) {
            notifyChange(uri)
        }
        return count
    }

    override fun update(uri: Uri, values: ContentValues?, where: String?, whereArgs: Array<String>?): Int {
        checkSignature(callingPackage)
        // TODO: 9/14/16 updates should be limited to very few things
        val db = dbHelper!!.writableDatabase
        val match = URI_MATCHER.match(uri)
        val count = db.update(getTableFromMatch(match), values, where, whereArgs)
        if (count > 0) {
            notifyChange(uri)
        }
        return count
    }

    private fun notifyChange(uri: Uri) {
        //crashing if context or contentResolver are null is appropriate here
        //noinspection ConstantConditions
        context!!.contentResolver.notifyChange(uri, null)
    }

    private fun  checkSignature(callingPackage: String) {
        communicationManager!!.checkSignature(callingPackage)
    }

    companion object {
        val URI_MATCHER: UriMatcher

        private val ADDRESS_LIST = 1
        private val ADDRESS_ID = 2
        private val TRANSACTION_LIST = 3
        private val TRANSACTION_ID = 4
        private val TRANSACTION_OUTPUT_LIST = 5
        private val TRANSACTION_OUTPUT_ID = 6

        init {
            URI_MATCHER = UriMatcher(UriMatcher.NO_MATCH)
            URI_MATCHER.addURI("*", Address.TABLE_NAME, ADDRESS_LIST)
            URI_MATCHER.addURI("*", Address.TABLE_NAME + "/#", ADDRESS_ID)
            URI_MATCHER.addURI("*", Transaction.TABLE_NAME, TRANSACTION_LIST)
            URI_MATCHER.addURI("*", Transaction.TABLE_NAME + "/#", TRANSACTION_ID)
            URI_MATCHER.addURI("*", TransactionOutput.TABLE_NAME, TRANSACTION_OUTPUT_LIST)
            URI_MATCHER.addURI("*", TransactionOutput.TABLE_NAME + "/#", TRANSACTION_OUTPUT_ID)
        }

        private fun getTableFromMatch(match: Int): String {
            return when (match) {
                ADDRESS_LIST, ADDRESS_ID -> Address.TABLE_NAME
                TRANSACTION_LIST, TRANSACTION_ID -> Transaction.TABLE_NAME
                TRANSACTION_OUTPUT_LIST, TRANSACTION_OUTPUT_ID -> TransactionOutput.TABLE_NAME
                else -> throw IllegalArgumentException("Unknown match " + match)
            }
        }
    }
}
